File: test1/t1-2/list.cpp
#include <iostream>
#include "list.h"

struct List
{
    ListElement *head;
};

struct ListElement
{
    int value;
    ListElement *next;
};

List *createList()
{
    List *newList = new List;
    newList->head = nullptr;
    return newList;
}

void addElementTo(List *list, int value, int position)
{    
    if (position == 0)
    {
        ListElement *newElement = new ListElement;
        newElement->value = value;
        newElement->next = list->head;
        list->head = newElement;
        return;
    }

    ListElement *newElement = new ListElement;
    ListElement *temp = list->head;
    newElement->value = value;

    for (int i = 1; i < position; i++)
    {
        temp = temp->next;
    }

    newElement->next = temp->next;
    temp->next = newElement;

}

bool isEmpty(List *list)
{
    return (list->head == nullptr);
}

int sizeList(List *list)
{
    int result = 0;
    ListElement *temp = list->head;
    while (temp != nullptr)
    {
        result++;
        temp = temp->next;
    }
    return result;
}

void printList(List *list)
{

    ListElement *temp = list->head;

    while (temp != nullptr)
    {
        std::cout << temp->value << " ";
        temp = temp->next;
    }
}

int peekListElement(List *list, int position)
{
    ListElement *temp = list->head;
    for (int i = 0; i < position; i++)
    {
        temp = temp->next;
    }

    return temp->value;
}

void deleteElementFrom(List *list, int position)
{
    if (position == 0)
    {
        ListElement *temp = list->head;
        list->head = temp->next;
        delete temp;
        return;
    }

    ListElement *temp = list->head;
    for (int i = 0; i < position - 1; i++)
    {
        temp = temp->next;
    }

    ListElement *deleteThis = temp->next;
    temp->next = temp->next->next;

    delete deleteThis;
}

void deleteList(List *&list)
{
    while (!isEmpty(list))
        deleteElementFrom(list, 0);

    delete list;
}

void clearList(List *list)
{
    while(!isEmpty(list))
        deleteElementFrom(list, 0);
}

File: sem2/p1-2/src/com/group144/kuzmin/List.java
package com.group144.kuzmin;

public class List {
    private ListElement head;

    public boolean isEmpty() {

        return this.head == null;
    }

    public int size() {
        ListElement temp = this.head;
        int result = 0;

        while (temp != null) {
            result++;
            temp = temp.next;
        }

        return result;
    }

    public void add(int value, int position) {
        if (position == 0) {
            this.head = new ListElement(value, this.head);
            return;
        }

        ListElement temp = this.head;

        for (int i = 1; i < position; i++)
            temp = temp.next;

        temp.next = new ListElement(value, temp.next);
    }

    public int peek(int position) {
        ListElement temp = this.head;

        for (int i = 0; i < position; i++)
            temp = temp.next;

        return temp.value;
    }

    public void delete(int position) {
        if (position == 0) {
            this.head = this.head.next;
            return;
        }

        ListElement temp  = this.head;

        for (int i = 1; i < position; i++)
            temp = temp.next;

        temp.next = temp.next.next;
    }

    public void print() {
        ListElement temp = this.head;
        int i = 0;

        while (temp != null) {
            System.out.println("Pos: " + i + " value: " + temp.value);
            temp = temp.next;
            i++;
        }
    }

    private class ListElement {
        private int value;
        private ListElement next;

        ListElement(int value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }
}